File: src/number.cc
#include "v8.h"
#include "node.h"
#include <math.h>
#include <stdint.h>

using namespace v8;

Handle<Value> Add(const Arguments &args);
Handle<Value> Substract(const Arguments &args);
Handle<Value> Multiply(const Arguments &args);
Handle<Value> Divide(const Arguments &args);
Handle<Value> Power(const Arguments &args);

void 
Initialize(Handle<Object> exports) {
    exports->Set(String::NewSymbol("value"),
            Number::New(666));

    exports->Set(String::NewSymbol("add"),
            FunctionTemplate::New(Add)->GetFunction());
    exports->Set(String::NewSymbol("substract"),
            FunctionTemplate::New(Substract)->GetFunction());
    exports->Set(String::NewSymbol("multiply"),
            FunctionTemplate::New(Multiply)->GetFunction());
    exports->Set(String::NewSymbol("divide"),
            FunctionTemplate::New(Divide)->GetFunction());
    exports->Set(String::NewSymbol("power"),
            FunctionTemplate::New(Power)->GetFunction());
}

Handle<Value>
Add(const Arguments &args) {
    HandleScope scope;

    if (!args[0]->IsNumber() || !args[1]->IsNumber()) {
        ThrowException(Exception::TypeError(
                    String::New("Arguments should be Numbers.")));

        return scope.Close(Undefined());
    }

    int64_t a = args[0]->NumberValue();
    int64_t b = args[1]->NumberValue();

    return scope.Close(Number::New(a + b));
}

Handle<Value>
Substract(const Arguments &args) {
    HandleScope scope;

    if (!args[0]->IsNumber() || !args[1]->IsNumber()) {
        ThrowException(Exception::TypeError(
                    String::New("Arguments should be Numbers.")));

        return scope.Close(Undefined());
    }

    int64_t a = args[0]->NumberValue();
    int64_t b = args[1]->NumberValue();

    return scope.Close(Number::New(a - b));
}

Handle<Value>
Multiply(const Arguments &args) {
    HandleScope scope;

    if (!args[0]->IsNumber() || !args[1]->IsNumber()) {
        ThrowException(Exception::TypeError(
                    String::New("Arguments should be Numbers.")));

        return scope.Close(Undefined());
    }

    int64_t a = args[0]->NumberValue();
    int64_t b = args[1]->NumberValue();

    return scope.Close(Number::New(a * b));
}

Handle<Value>
Divide(const Arguments &args) {
    HandleScope scope;

    if (!args[0]->IsNumber() || !args[1]->IsNumber()) {
        ThrowException(Exception::TypeError(
                    String::New("Arguments should be Numbers.")));

        return scope.Close(Undefined());
    }

    int64_t a = args[0]->NumberValue();
    int64_t b = args[1]->NumberValue();

    return scope.Close(Number::New(a / b));
}

Handle<Value>
Power(const Arguments &args) {
    HandleScope scope;

    if (!args[0]->IsNumber() || !args[1]->IsNumber()) {
        ThrowException(Exception::TypeError(
                    String::New("Arguments should be Numbers.")));

        return scope.Close(Undefined());
    }

    int64_t a = args[0]->NumberValue();
    int64_t b = args[1]->NumberValue();

    return scope.Close(Number::New(pow(a, b)));
}

NODE_MODULE(number, Initialize)

File: test/number.js
var chai   = require('chai');
var addons = require('../lib');

describe('Number', function() {

  describe('value', function() {

    it('should be 666', function() {
      chai.expect(addons.number.value).to.equal(666);
    });

  });

  describe('#add', function() {

    it('should return 3000', function() {
      chai.expect(addons.number.add(2000, 1000)).to.equal(3000);
    });

  });

  describe('#substract', function() {

    it('should return 1000', function() {
      chai.expect(addons.number.substract(2000, 1000)).to.equal(1000);
    });

  });

  describe('#multiply', function() {

    it('should return 81', function() {
      chai.expect(addons.number.multiply(9, 9)).to.equal(81);
    });

  });

  describe('#divide', function() {

    it('should return 0xf', function() {
      chai.expect(addons.number.divide(0xfff, 0xff)).to.equal(0x10);
    });

  });

  describe('#power', function() {

    it('should return 256', function() {
      chai.expect(addons.number.power(2, 8)).to.equal(256);
    });

  });

});
